Added warning msg for `kubectl get`

vefimova opened the pull request:
- added warning description regarding terminated pods to `get` long help message
- added printing of warning message in case of `get pods` if there are hidden pods

Fixes #22986  (initiall PR and discussion are here #26417)
## **Output examples:**
### # kubectl get pods

```
NAME                       READY     STATUS             RESTARTS   AGE
dapi-test-pod1             0/1       Terminating        0          22h
liveness-http              0/1       CrashLoopBackOff   11245      22d
ubuntu1-1206318548-oh9tc   0/1       CrashLoopBackOff   2336       8d
  info: 1 completed object(s) was(were) not shown in pods list. Pass --show-all to see all objects.

```
### #  kubectl get pods,namespaces

```
NAME                          READY     STATUS             RESTARTS   AGE
po/dapi-test-pod1             0/1       Terminating        0          22h
po/liveness-http              1/1       Running            11242      22d
po/ubuntu1-1206318548-oh9tc   0/1       CrashLoopBackOff   2335       8d
 info: 1 completed object(s) was(were) not shown in pods list. Pass --show-all to see all objects.

NAME             STATUS    AGE
ns/default       Active    89d
ns/kube-system   Active    41d
```
### # kubectl get pods -a

```
NAME                       READY     STATUS             RESTARTS   AGE
busybox                    0/1       Error              0          27d
dapi-test-pod1             0/1       Terminating        0          22h
liveness-http              0/1       CrashLoopBackOff   11245      22d
ubuntu1-1206318548-oh9tc   0/1       CrashLoopBackOff   2336       8d
```
### # kubectl get -h

```
Display one or many resources.

Possible resource types include (case insensitive): pods (aka 'po'), services (aka 'svc'), deployments (aka 'deploy'),
replicasets (aka 'rs'), replicationcontrollers (aka 'rc'), nodes (aka 'no'), events (aka 'ev'), limitranges (aka 'limits'),
persistentvolumes (aka 'pv'), persistentvolumeclaims (aka 'pvc'), resourcequotas (aka 'quota'), namespaces (aka 'ns'),
serviceaccounts (aka 'sa'), ingresses (aka 'ing'), horizontalpodautoscalers (aka 'hpa'), daemonsets (aka 'ds'), configmaps (aka 'cm'),
componentstatuses (aka 'cs), endpoints (aka 'ep'), petsets (alpha feature, may be unstable) and secrets.

This command will hide resources that have completed. For instance, pods that are in the Succeeded or Failed phases.
You can see the full results for any resource by providing the '--show-all' flag.

By specifying the output as 'template' and providing a Go template as the value
of the --template flag, you can filter the attributes of the fetched resource(s).

Examples:

.........
```

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/28352)

<!-- Reviewable:end -->

jlowdermilk commented on pkg/kubectl/kubectl.go:
```diff
@@ -30,6 +30,11 @@ import (
 
 const (
 	kubectlAnnotationPrefix = "kubectl.kubernetes.io/"
+	Warn                    = `
```
Give this a more meaningful name, e.g. `HiddenPodsWarning`.

vefimova replied:
done

jlowdermilk commented on pkg/kubectl/resource_printer.go:
```diff
@@ -145,6 +145,9 @@ type ResourcePrinter interface {
 	HandledResources() []string
+	//Can be used to print out warning/clarifications if needed
+	//after all objects were printed
+	FinishPrint(io.Writer) error
```
Since this is only ever used by `HumanReadablePrinter`, it should not be added to the interface. Suggest doing the check for hidden pods at the end of [`HumanReadablePrinter.PrintObj`](https://github.com/vefimova/kubernetes/blob/ec2a786d826ca6d97d5cf4f856a08bcdfc1313a0/pkg/kubectl/resource_printer.go#L2049), and printing the warning there.

vefimova replied:
That can't be done so because in case of printing pods one by one using printPod handler, so it's needed generic approach here as we don't know how much pods should be printed and whether any of them will be hidden or not.

jlowdermilk replied:
`PrintObj` is what calls `printPodList`, so if you put the warning message in there before returning (maybe use a defer?), it should give you the same output as
```
err := printer.PrintObj(e.Object, out)
if err == nil {
		printer.FinishPrint(out)
}
```
which is what you're doing now.

vefimova replied:
In case of get pods without watch option set, looping on Pod list is done in RunGet function itself and PrintObj func is called to print pods one by one using printPod handler.

jlowdermilk commented on pkg/kubectl/resource_printer.go:
```diff
@@ -452,8 +510,8 @@ var networkPolicyColumns = []string{"NAME", "POD-SELECTOR", "AGE"}
-	h.Handler(podColumns, printPod)
-	h.Handler(podColumns, printPodList)
+	h.Handler(podColumns, printPod, &h.hiddenPodsNum)
+	h.Handler(podColumns, printPodList, &h.hiddenPodsNum)
```
Rather than passing in a pointer to `h.hiddenPodsNum`, try using a wrapper for the printFunc.
Could be either a private member of h or an anonymous function, something like
```
 func (h *HumanReadablePrinter) addDefaultHandlers() {
	h.Handler(podColumns, func(pod *api.Pod, w io.Writer, options PrintOptions) {
		reason := string(pod.Status.Phase)
		// if not printing all pods, skip terminated pods (default)
 		if !options.ShowAll && (reason == string(api.PodSucceeded) || reason == string(api.PodFailed)) {
			h.hiddenPods++
 			return nil
		}
		return printPod(pod, w, options)
	})
```

That way you don't have to change the interface or deal with reflection hacks.

vefimova replied:
Yes, thanks, agree, that's much much better! Will fix it in that way.

jlowdermilk commented on pkg/kubectl/resource_printer.go:
```diff
@@ -352,23 +377,48 @@ func NewHumanReadablePrinter(noHeaders, withNamespace bool, wide bool, showAll b
 // Handler adds a print handler with a given set of columns to HumanReadablePrinter instance.
 // See validatePrintHandlerFunc for required method signature.
-func (h *HumanReadablePrinter) Handler(columns []string, printFunc interface{}) error {
+func (h *HumanReadablePrinter) Handler(columns []string, printFuncsArgs ...interface{}) error {
```
This has serious code smell. Instead of hacking the handler interface to pass arbitrary args, suggest using a wrapper function around `printPod`, `printPodList` as the default handler. See comment in `addDefaultHandlers()`.

jlowdermilk commented on pkg/kubectl/resource_printer.go:
```diff
@@ -412,6 +462,14 @@ func (h *HumanReadablePrinter) HandledResources() []string {
 
+func (h *HumanReadablePrinter) FinishPrint(output io.Writer) error {
+	if !h.options.NoHeaders && !h.options.ShowAll && h.hiddenPodsNum > 0 {
+		_, err := fmt.Fprintf(output, "\n *Be aware of %d pod(s) is(are) in terminated state and hidden from the displayed list!\n", h.hiddenPodsNum)
```
Wherever this print statement ends up going, I think the warning should be printed to Stderr, not output. You can use a private `io.Writer` member on `HumanReadablePrinter` to store `os.Stderr` and change it for testing.

therc replied:
If this logic is going to be reused for other objects as in #27510, this message needs to be parameterized somehow. Either the caller passes a custom message or `HumanReadablePrinter` takes care of it in one of two ways:

1. it keeps a mapping of object type -> message
2. it uses a more generic message (i.e. does not mention terminated state). It can either take `%d pod(s)` in the current version that you are suggesting and replace it with `%d object(s)` or it can inject the object type as an additional formatting placeholder, i.e. `%d %s(s)`.

smarterclayton commented:
I would like to review this as well, but will not have time until Monday.

therc commented on pkg/kubectl/resource_printer.go:
```diff
@@ -334,9 +358,10 @@ type PrintOptions struct {
+	handlerMap    map[reflect.Type]*handlerEntry
+	options       PrintOptions
+	lastType      reflect.Type
+	hiddenPodsNum int
```
We need to do something similar for replicasets in #27510 and we might need it again for other objects. Can you change the field name to something more generic like `hiddenEntryNum`, `hiddenObjNum`, `hiddenCount` or similar? That way, we can reuse the code and logic.

vefimova replied:
Added changes for reuse the same var `hiddenObjNum` for all objects and valid printing in case of user getting several resource types at once: kubectl get pods,namespaces (added output example in PR description)

therc replied:
Thank you so much!

soltysh commented on docs/proposals/container-init.md:
```diff
@@ -18,11 +18,6 @@
 If you are using a released version of Kubernetes, you should
 refer to the docs that go with that version.
 
-<!-- TAG RELEASE_LINK, added by the munger automatically -->
```
Was that intentional?

vefimova replied:
This change was autogenerated after running pre-commit hook. Not sure why it's appearing.

vefimova committed:
Added warning msg for `kubectl get`
  - added warning description regarding terminated objects to `get` long help message
  - added printing of warning message in case of `get pods` if there are hidden pods
Fixes #22986

smarterclayton commented:
Looks better, will review.

jlowdermilk commented:
@smarterclayton still planning to review this?

smarterclayton commented on pkg/kubectl/kubectl.go:
```diff
@@ -30,6 +30,11 @@ import (
 const (
 	kubectlAnnotationPrefix = "kubectl.kubernetes.io/"
+	HiddenPodsWarning       = `
+WARN: By default only active pods are displayed, to display terminated
```
This message needs to be more generic, and doesn't need the warning prefix.  It really belongs directly inside of the get comment.

```
This command will hide by default some resources that are being terminated, like pods. You can see the full results for any resource by passing the `--show-all` flag.
```

Should be a separate paragraph and in get.go in the help text.

smarterclayton replied:
A separate warning can remain here, but I would recommend making it shorter:

```
WARNING: Some completed pods were not displayed. Pass --show-all to display all pods.
```

vefimova replied:
fixed

smarterclayton commented on pkg/kubectl/resource_printer.go:
```diff
@@ -407,6 +433,14 @@ func (h *HumanReadablePrinter) HandledResources() []string {
 
+func (h *HumanReadablePrinter) FinishPrint(output io.Writer, res string) error {
+	if !h.Options.NoHeaders && !h.Options.ShowAll && h.hiddenObjNum > 0 {
+		_, err := fmt.Fprintf(output, "  *Be aware of %d object(s) is(are) hidden from the displayed %s list!\n\n", h.hiddenObjNum, res)
```
I recommend wording this like:

```
info: %d completed pods were not shown. Pass --show-all to see all pods.
```

vefimova replied:
fixed

smarterclayton commented on pkg/kubectl/cmd/get.go:
```diff
@@ -355,6 +364,10 @@ func RunGet(f *cmdutil.Factory, out io.Writer, cmd *cobra.Command, args []string
 		if printer == nil || lastMapping == nil || mapping == nil || mapping.Resource != lastMapping.Resource {
+			if printer != nil {
+				w.Flush()
+				printer.FinishPrint(os.Stderr, lastMapping.Resource)
```
Why is this here if you have finish print?

vefimova replied:
It's for the case user requested to get several resource type at once. So after one resource type list is ended we should check whether there's something to add in the end (look at "kubectl get pods,namespaces" in output examples above in PR description).

smarterclayton replied:
Ok, thanks

smarterclayton commented on pkg/kubectl/cmd/get.go:
```diff
@@ -382,5 +395,9 @@ func RunGet(f *cmdutil.Factory, out io.Writer, cmd *cobra.Command, args []string
 	}
+	w.Flush()
+	if printer != nil {
+		printer.FinishPrint(os.Stderr, lastMapping.Resource)
```
You should not use `os.Stderr` directly, but instead get an `errout` io.Writer passed in constructor by the code that initializes the command.

vefimova replied:
fixed

vefimova commented:
@jlowdermilk @smarterclayton 

So guys, WDYT is patch ready to be merged?

smarterclayton commented:
LGTM

smarterclayton commented:
As a future note I recommended to someone else with a similar problem that the next change to this code should hoist filtering up as a generic tool that happens before printing.  That will make this more resilient to future change

therc commented:
That would be me for `get rs`.

AdoHe commented:
@smarterclayton I think you mean this #29115, actually I am planning this.
